### ISSUE TITLE
useKeyPress doesn't track repeated combination keypresses until the entire combination is released

### ISSUE THREAD

**Issue by adwilk** [wbkd/react-flow#2248]
### Describe the Bug

When using `useKeyPress` to track combinations of keys, for example `Meta+z`, repeated keypresses are not captured if you continue to hold down the `Meta` key.

### Your Example Website or App

https://codesandbox.io/s/recursing-driscoll-stvwou?file=/src/App.js

### Steps to Reproduce the Bug or Issue

When `const comboPressed = useKeyPress("Meta+z");`:
- Press and hold `Meta`
- Press and hold `z`
- `comboPressed` is true
- Release `z`
- `comboPressed` is still true and so additional presses of `z` can't be detected
- Release `Meta`
- `comboPressed` is now false

The inverse does not have the same behaviour:
- Press and hold `z`
- Press and hold `Meta`
- `comboPressed` is true
- Release `Meta`
- `comboPressed` is now _false_ and so additional presses of `z` _can_ be detected

### Expected behavior

As a user I would expect that after I press `Meta and z`, and then continue to hold `Meta` but release `z` that the pressed state of `useKeyPress("Meta+z")` is false.

### Platform

- OS: macOS
- Browser: Chrome
- Version: 103.0.5060.53

**Comment by moklick** [comment]
It seems that on macOs, the keyup event doesn't get fired in combination with `'Meta'`. Need to look deeper into this and see how we can fix this.

**Comment by joeyballentine** [comment]
> It seems that on macOs, the keyup event doesn't get fired in combination with `'Meta'`. Need to look deeper into this and see how we can fix this.

It's not just macos. It happens to me on windows as well

**Comment by moklick** [comment]
Ok, thanks. We will need to check how other libs handle this

**Comment by allenhwkim** [comment]
Repeating keypress works with other keys, e.g. a, b. c, d, ..., z
However when it's conflict with the browser default shortcuts, it does not work e.g. Meta + z
I think it needs to have event.preventDefault()
<img width="413" alt="image" src="https://user-images.githubusercontent.com/1437734/202879853-fdbc0a0f-3ea8-49bb-b8aa-f4de195feb44.png">

**Comment by allenhwkim** [comment]
IMO, it should use `keydown` event instead of `keypress` event.
The following code is what I have tried without `useKeyPress` [API](https://reactflow.dev/docs/api/hooks/use-key-press/), and it seems working 

Note. this example is using [state management library](https://reactflow.dev/docs/guides/state-management/)
```
import * as React from 'react';
import {KeyboardEvent} from 'react';
import ReactFlow from 'reactflow';
import 'reactflow/dist/style.css';

import useStore from './store';

function Flow() {
  const { nodes, edges, onNodesChange, onEdgesChange, onConnect } = useStore();

  const onKeyDown = (event : KeyboardEvent) => {
    const ctrl = event.ctrlKey ? 'Control-' : '';
    const alt = event.altKey ? 'Alt-' : '';
    const meta = event.metaKey ? 'Meta-' : '';
    const shift = event.shiftKey ? 'Shift-' : '';
    const key = `${ctrl}${alt}${shift}${meta}${event.key}`;
    if (!['Alt', 'Shift', 'Meta', 'Control'].includes(event.key)) {
      console.log(`${key} pressed`);
    }
  }

  return (
    <ReactFlow 
      tabIndex={0}
      onKeyDown={(e) => onKeyDown(e)}
      nodes={nodes}
      edges={edges}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      onConnect={onConnect}
    ></ReactFlow>
  );
}

export default Flow;
```

**Comment by Max-ChenFei** [comment]
For example, we usekeypress hook 
```const copy = useKeyPress('Control+c');```

From the code 
https://github.com/wbkd/react-flow/blob/f79620f4c0c1ab96eb01986988e50eb5274f9d54/packages/core/src/hooks/useKeyPress.ts#L70-L72
Here even if we just release one key (for example `c` here) for combination keys, all pressedkeys (`control`, `c`) will be clear so that another held key (`Control`) is not memorized, so if we pressed again, the pressedkeys will be only `c`, so it will not match `Control+c`.

A possible solution I tested is just to delete the key that was release.
 `pressedKeys.current.delete(event[keyOrCode]);` at line 72.

**Comment by Vochsel** [comment]
Just chiming back in here to say @Max-ChenFei 's answer _seems_ to work.

```
if (isMatchingKey(keyCodes, pressedKeys.current, true)) {
    setKeyPressed(false);
}
                
pressedKeys.current.delete(event[keyOrCode]);
```

Moving out the delete keyCode and just setting key pressed to false seems to work.

**Comment by Max-ChenFei** [comment]
Hi, @Vochsel  just a temp solution, I add `onKeyDown` event to the parent of the `ReatFlow` and use the default keydown event. 

```
 <div
        ref={sceneDomRef}
        tabIndex={0}
        onKeyDown={(e) => {
          if (
            e.target === sceneDomRef.current ||
            (e.target as HTMLElement).classList.contains('react-flow__node')
          )
            onKeyDown(e, sceneState ?? undefined, sceneDomRef);
        }}
        style={{ outline: 'none' }}
      >
     
        <ReactFlow
....
```

**Comment by moklick** [comment]
thanks, I will check it.

**Comment by Vochsel** [comment]
Chiming back in here to say I'm sometimes hitting an edge cases where no key press works till the blur reset handler is refired...

**Comment by moklick** [comment]
Yep, this doesn't work unfortunately. For the explained case it works better, but it introduces a new bug :/

**Comment by joeyballentine** [comment]
@moklick in chaiNNer we use this package: https://www.npmjs.com/package/react-hotkeys-hook and it works quite well

**Comment by Iaotle** [comment]
@joeyballentine this fixed my issue as well. @moklick I suggest a move to this handler, as it is more feature-rich and does not interfere unlike the one in react-flow. Changing useShortcut from the copy-paste example to:

```ts
function useShortcut(keyCode, callback: Function, active: boolean | undefined): void {
	useHotkeys(keyCode, () => callback(), { enabled: active }, [active, callback]);
}
```
solved my issue where copy-paste commands that were not meant to be intercepted at all would trigger only on 2nd or 3rd attempt.

**Comment by vincent067** [comment]
Hey, just ran into this exact issue today! 😅

I am building a diagram editor where users need to duplicate nodes with Ctrl+D multiple times quickly. The second Ctrl+D never fires if they do not release Ctrl completely between presses.

I dug into the source a bit - looks like useKeyPress uses keydown/keyup events and tracks pressedKeys in a Set. The problem is when you keep Ctrl held down, the second D keydown fires but the hook still thinks the combo is already active from the first press.

My temporary hack is to clear the pressedKeys Set after a small delay when a combo is triggered, but that feels wrong and might cause race conditions.

Has anyone found a cleaner workaround? Or is this something the team would accept a PR for? Happy to take a stab at it if there is interest.

For reference, I am on @xyflow/react 12.0.3.
